All imports

In [1]:
import warnings
warnings.filterwarnings('ignore')

import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score 
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import RobustScaler
from sklearn.preprocessing import Normalizer
from sklearn.model_selection import GridSearchCV
from sklearn.neural_network import MLPClassifier 

Question 1:
Load the data set into Python using, e.g., load_wine or genfromtxt, as appropriate. In the case of the USPS dataset, merge the original training and test sets into one dataset.

In [2]:
from sklearn.datasets import load_wine
wine=load_wine()

In [3]:
USPS_test=np.genfromtxt("zip.test")
USPS_train=np.genfromtxt("zip.train")
USPS=np.concatenate((USPS_train,USPS_test),axis=0)

Question 2: Divide the dataset into a training set and a test set. You may use the function train_test_split. Use your birthday in the format DDMM as random_state (omit leading zeros if any).

-> wine dataset

In [4]:
X_wine=wine.data
y_wine=wine.target
X_train_wine, X_test_wine, y_train_wine, y_test_wine=train_test_split(X_wine,y_wine,random_state=206)

-> USPS Dataset

In [5]:
X_USPS=USPS[:,:-1]
y_USPS=USPS[:,-1].astype(int)
X_train_USPS, X_test_USPS, y_train_USPS, y_test_USPS=train_test_split(X_USPS,y_USPS,random_state=206)

Question 3: Using cross-validation and the training set only, estimate the generalization accuracy of the SVM with the default values of the parameters. You may use the function cross_val_score.

-> Wine Dataset

In [6]:
svm = SVC()
scores_wine=cross_val_score(svm,X_train_wine,y_train_wine)
accuracy_wine_train = np.mean(scores_wine)
print('Accuracy (training set) of the SVM on winedata set with default values of parameters:',accuracy_wine_train)

Accuracy (training set) of the SVM on winedata set with default values of parameters: 0.4289073750651102


-> USPS Dataset

In [7]:
scores_USPS=cross_val_score(svm,X_train_USPS,y_train_USPS)
accuracy_USPS_train = np.mean(scores_USPS)
print('Accuracy (training set) of the SVM on USPS dataset with default values of parameters:',accuracy_USPS_train)

Accuracy (training set) of the SVM on USPS dataset with default values of parameters: 0.9886705655186029


Question 4: Find the test error rate of the SVM with the default values of parameters, compare it with the estimate obtained in the previous task (task 3), and write your observations in a markdown cell of your Jupyter notebook.

-> Wine Dataset

In [8]:
svc_wine=SVC()
svc_wine.fit(X_train_wine,y_train_wine)
test_error_rate_wine=1-svc_wine.score(X_test_wine,y_test_wine)
print('Test Error Rate of the SVM:',test_error_rate_wine)
print('Accuracy on training set:',accuracy_wine_train)

Test Error Rate of the SVM: 0.6
Accuracy on training set: 0.4289073750651102


-> USPS Dataset

In [9]:
svc_USPS=SVC()
svc_USPS.fit(X_train_USPS,y_train_USPS)
test_error_rate_USPS=1-svc_USPS.score(X_test_USPS,y_test_USPS)
print('Test Error Rate of the SVM:',test_error_rate_USPS)
print('Accuracy on training set:',accuracy_USPS_train)

Test Error Rate of the SVM: 0.010322580645161339
Accuracy on training set: 0.9886705655186029


Observation: It was observed that the result obtained for Accuracy on training set and test error rate on test dataset for both the datasets (Wine and USPS) are almost similar meaning that the model is underfitting where training and test performance is quite similar. The model can be improved by increasing the value of C and gamma.

Question 5: Create a pipeline for SVM involving data normalization and SVC, and use grid search and cross-validation to tune parameters C and gamma for the pipeline, avoiding data snooping and data leakage.

-> Wine Dataset

In [10]:
import warnings
warnings.filterwarnings('ignore')

param_grid = {'svc__C': [0.01, 0.1, 1, 10, 100],'svc__gamma': [0.001, 0.01, 0.1, 1, 10, 100]}

#---- MinMaxScaler -------
pipe_wine_MinMaxScaler = make_pipeline(MinMaxScaler(), SVC())
grid_MinMaxScaler = GridSearchCV(pipe_wine_MinMaxScaler, param_grid=param_grid)
grid_MinMaxScaler.fit(X_train_wine,y_train_wine)
print("Best cross-validation accuracy on training dataset (MinMaxScaler):", grid_MinMaxScaler.best_score_) 

#----- Standard Scaler --------
pipe_wine_StandardScaler = make_pipeline(StandardScaler(), SVC())
grid_StandardScaler = GridSearchCV(pipe_wine_StandardScaler, param_grid=param_grid)
grid_StandardScaler.fit(X_train_wine,y_train_wine)
print("Best cross-validation accuracy on training dataset(StandardScaler):", grid_StandardScaler.best_score_) 

#------ RobustScaler ----------
pipe_wine_RobustScaler = make_pipeline(RobustScaler(), SVC())
grid_RobustScaler = GridSearchCV(pipe_wine_RobustScaler, param_grid=param_grid)
grid_RobustScaler.fit(X_train_wine,y_train_wine)
print("Best cross-validation accuracy on training dataset (RobustScaler):", grid_RobustScaler.best_score_) 

#----- Normalizer ------------
pipe_wine_Normalizer = make_pipeline(Normalizer(), SVC())
grid_Normalizer = GridSearchCV(pipe_wine_Normalizer, param_grid=param_grid) 
grid_Normalizer.fit(X_train_wine,y_train_wine)
print("Best cross-validation accuracy on training dataset(Normalizer):", grid_Normalizer.best_score_) 


Best cross-validation accuracy on training dataset (MinMaxScaler): 0.9924812030075187
Best cross-validation accuracy on training dataset(StandardScaler): 0.9924812030075187
Best cross-validation accuracy on training dataset (RobustScaler): 0.9849624060150376
Best cross-validation accuracy on training dataset(Normalizer): 0.9097744360902256


-> USPS Dataset

In [11]:
param_grid = {'svc__C': [0.01, 1, 10],'svc__gamma': [0.001, 0.01, 0.1]}

#---- MinMaxScaler -------
pipe_USPS_MinMaxScaler = make_pipeline(MinMaxScaler(), SVC())
grid_MinMaxScaler_USPS = GridSearchCV(pipe_USPS_MinMaxScaler, param_grid=param_grid)
grid_MinMaxScaler_USPS.fit(X_train_USPS,y_train_USPS)
print("Best cross-validation accuracy on training dataset (MinMaxScaler):", grid_MinMaxScaler_USPS.best_score_)  

#----- Standard Scaler --------
pipe_USPS_StandardScaler = make_pipeline(StandardScaler(), SVC())
grid_StandardScaler_USPS = GridSearchCV(pipe_USPS_StandardScaler, param_grid=param_grid)
grid_StandardScaler_USPS.fit(X_train_USPS,y_train_USPS)
print("Best cross-validation accuracy on training dataset(StandardScaler):", grid_StandardScaler_USPS.best_score_) 

#------ RobustScaler ----------
pipe_USPS_RobustScaler = make_pipeline(RobustScaler(), SVC())
grid_RobustScaler_USPS = GridSearchCV(pipe_wine_RobustScaler, param_grid=param_grid)
grid_RobustScaler_USPS.fit(X_train_USPS,y_train_USPS)
print("Best cross-validation accuracy on training dataset (RobustScaler):", grid_RobustScaler_USPS.best_score_) 

#----- Normalizer ------------
pipe_USPS_Normalizer = make_pipeline(Normalizer(), SVC())
grid_Normalizer_USPS = GridSearchCV(pipe_USPS_Normalizer, param_grid=param_grid) 
grid_Normalizer_USPS.fit(X_train_USPS,y_train_USPS)
print("Best cross-validation accuracy on training dataset(Normalizer):", grid_Normalizer_USPS.best_score_) 


Best cross-validation accuracy on training dataset (MinMaxScaler): 0.9939767675319088
Best cross-validation accuracy on training dataset(StandardScaler): 0.9954108705005019
Best cross-validation accuracy on training dataset (RobustScaler): 0.9870930732826617
Best cross-validation accuracy on training dataset(Normalizer): 0.9929728954538936


Question 6: Fit the GridSearchCV object of task 5 to the training set and use it to predict the test labels. Write the resulting test accuracy in your Jupyter notebook.

-> Wine Dataset

In [12]:
grid_MinMaxScaler.fit(X_train_wine,y_train_wine)
pred_wine_MinMaxScaler=grid_MinMaxScaler.predict(X_test_wine)
print("Test set score (MinMaxScaler):", np.mean(y_test_wine==pred_wine_MinMaxScaler))

grid_StandardScaler.fit(X_train_wine,y_train_wine)
pred_wine_StandardScaler=grid_StandardScaler.predict(X_test_wine)
print("Test set score (StandardScaler):", np.mean(y_test_wine==pred_wine_StandardScaler)) 

grid_RobustScaler.fit(X_train_wine,y_train_wine)
pred_wine_RobustScaler=grid_RobustScaler.predict(X_test_wine)
print("Test set score (RobustScaler):",np.mean(y_test_wine==pred_wine_RobustScaler)) 

grid_Normalizer.fit(X_train_wine,y_train_wine)
pred_wine_Normalizer=grid_Normalizer.predict(X_test_wine)
print("Test set score (Normalizer):",np.mean(y_test_wine==pred_wine_Normalizer))

Test set score (MinMaxScaler): 0.9777777777777777
Test set score (StandardScaler): 1.0
Test set score (RobustScaler): 0.9777777777777777
Test set score (Normalizer): 0.9777777777777777


Normalization techiques used on Wine Data set were StandardScaler, MinMaxScaler, RobustScaler, and Normalizer.

After training the dataset and testing it on different normalisation techniques, it was observed that for:
1. MinMaxScaler: 0.9924812030075187 (training) and 0.9777777777777777 (test score) - The performance decreased.
2. StandardScaler: 0.9924812030075187 (training) and 1.0 (test score) - The model overfitted
3. RobustScaler: 0.9849624060150376 (training) and 0.9777777777777777 (test score) - The model performace decreased
4. Normalizer: 0.9097744360902256 (training) and 0.9777777777777777 (test score) - The performace of the model increased.

Hence, for wine dataset Normalizer is a better normalization technique.

-> USPS Dataset

In [13]:
grid_MinMaxScaler_USPS.fit(X_train_USPS,y_train_USPS)
pred_USPS_MinMaxScaler=grid_MinMaxScaler_USPS.predict(X_test_USPS)
print("Test set score (MinMaxScaler):", np.mean(y_test_USPS==pred_USPS_MinMaxScaler))

grid_StandardScaler_USPS.fit(X_train_USPS,y_train_USPS)
pred_USPS_StandardScaler=grid_StandardScaler_USPS.predict(X_test_USPS)
print("Test set score (StandardScaler):", np.mean(y_test_USPS==pred_USPS_StandardScaler)) 

grid_RobustScaler_USPS.fit(X_train_USPS,y_train_USPS)
pred_USPS_RobustScaler=grid_RobustScaler_USPS.predict(X_test_USPS)
print("Test set score (RobustScaler):",np.mean(y_test_USPS==pred_USPS_RobustScaler)) 

grid_Normalizer_USPS.fit(X_train_USPS,y_train_USPS)
pred_USPS_Normalizer=grid_Normalizer_USPS.predict(X_test_USPS)
print("Test set score (Normalizer):",np.mean(y_test_USPS==pred_USPS_Normalizer))


Test set score (MinMaxScaler): 0.993978494623656
Test set score (StandardScaler): 0.9969892473118279
Test set score (RobustScaler): 0.9862365591397849
Test set score (Normalizer): 0.9931182795698925


Normalization techiques used on USPS Data set were StandardScaler, MinMaxScaler, RobustScaler, and Normalizer. 

After training the dataset and testing it on different normalisation techniques, it was observed that for:

1. MinMaxScaler: 0.9939767675319088 (training) and 0.993978494623656 (test score) - The performance is underfitting.
2. StandardScaler: 0.9954108705005019 (training) and 0.9969892473118279 (test score) - The model performance increased.
3. RobustScaler: 0.9870930732826617 (training) and 0.9862365591397849 (test score) - The model performace decreased
4. Normalizer: 0.9929728954538936 (training) and 0.9931182795698925 (test score) - The performace of the model is almost the same for training and test - underfitting.

Hence, for USPS dataset StandardScaler is a better normalization technique.

Question 7: Implement a cross-conformal predictor. You may use the KFold class for splitting into folds (start from 5 or 10 folds). For computing the conformity scores for each fold, you may use one of the GridSearchCV objects that you created in task 5 in combination with the decision_function method (see Section 3 of Lab Worksheet 9 for examples). Run your cross-conformal predictor on the two datasets, training it on the training set and testing on the test set.

Question 8: An alternative to implementing a cross-conformal predictor is to experi- ment with a neural network. Perform tasks 3–6 for the scikit-learn class MLPClassifier, as described in Lab Worksheet 8, Section 1. It is up to you to decide which parameters to fit using grid search.

-> Wine Dataset

In [14]:
mlp_wine=MLPClassifier(random_state=206)
scores_wine_mlp=cross_val_score(mlp_wine,X_train_wine,y_train_wine)
accuracy_wine_mlp = np.mean(scores_wine_mlp)
print('Generalization accuracy (train) with MLP Classifier:',accuracy_wine_mlp)
print('Generalizaton accuracy (training set) of the SVM on wine dataset (default parameters):',accuracy_wine_train)

mlp_wine.fit(X_train_wine,y_train_wine)
pred_wine_mlp=mlp_wine.predict(X_test_wine)
print('Test Error Rate (default parameters) with MLP Classifier:',np.mean(y_test_wine!=pred_wine_mlp))
print('Test Error Rate of the SVM:',test_error_rate_wine)

print('\n')
param_grid = {'mlpclassifier__solver': ['lbfgs', 'sgd', 'adam'],
              'mlpclassifier__hidden_layer_sizes': [[10],[10,10],[10,10,10]],
              'mlpclassifier__activation':['relu','tanh']}

#---- MinMaxScaler -------
pipe_wine_mlp_MinMaxScaler = make_pipeline(MinMaxScaler(), mlp_wine)
grid_MinMaxScaler_mlp = GridSearchCV(pipe_wine_mlp_MinMaxScaler, param_grid=param_grid)
grid_MinMaxScaler_mlp.fit(X_train_wine,y_train_wine)
pred_wine_MinMaxScaler_mlp=grid_MinMaxScaler_mlp.predict(X_test_wine)
print("Test Error rate (MinMaxScaler) for classifier:", np.mean(y_test_wine!=pred_wine_MinMaxScaler_mlp))
print("Test Error rate (MinMaxScaler) for wine dataset (SVC):", np.mean(y_test_wine!=pred_wine_MinMaxScaler))
print('\n')

#----- Normalizer ------------
pipe_wine_Normalizer_mlp = make_pipeline(Normalizer(), mlp_wine)
grid_Normalizer_mlp = GridSearchCV(pipe_wine_Normalizer_mlp, param_grid=param_grid) 
grid_Normalizer_mlp.fit(X_train_wine,y_train_wine)
pred_wine_Normalizer_mlp=grid_MinMaxScaler_mlp.predict(X_test_wine)
print("Test Error rate (Normalizer) for classifier on wine dataset:", 
      np.mean(y_test_wine!=pred_wine_Normalizer_mlp))
print("Test Error rate (Normalizer) for wine dataset:",np.mean(y_test_wine!=pred_wine_Normalizer))

Generalization accuracy (train) with MLP Classifier: 0.8723304838067224
Generalizaton accuracy (training set) of the SVM on wine dataset (default parameters): 0.4289073750651102
Test Error Rate (default parameters) with MLP Classifier: 0.13333333333333333
Test Error Rate of the SVM: 0.4


Test Error rate (MinMaxScaler) for classifier: 0.0
Test Error rate (MinMaxScaler) for wine dataset (SVC): 0.022222222222222223


Test Error rate (Normalizer) for classifier on wine dataset: 0.0
Test Error rate (Normalizer) for wine dataset: 0.022222222222222223


-> USPS Dataset

In [17]:
mlp_USPS=MLPClassifier(random_state=206)
scores_USPS_mlp=cross_val_score(mlp_USPS,X_train_USPS,y_train_USPS)
accuracy_USPS_mlp = np.mean(scores_USPS_mlp)
print('Generalization accuracy (train) with MLP Classifier:',accuracy_USPS_mlp)
print('Accuracy (training set) of the SVM on USPS dataset (default parameters):',accuracy_USPS_train)

mlp_USPS.fit(X_train_USPS,y_train_USPS)
test_error_rate_USPS_mlp=mlp_USPS.score(X_test_USPS,y_test_USPS)
print('Test Error Rate (default parameters) with MLP Classifier:',1-test_error_rate_USPS_mlp)
print('Test Error Rate of the SVM:',test_error_rate_USPS)

print('\n')
param_grid = {'mlpclassifier__solver': ['lbfgs', 'sgd', 'adam'],
              'mlpclassifier__hidden_layer_sizes': [[10],[10,10],[10,10,10]],
              'mlpclassifier__activation':['relu','tanh']}

#---- MinMaxScaler -------
pipe_USPS_mlp_MinMaxScaler = make_pipeline(MinMaxScaler(), mlp_USPS)
grid_MinMaxScaler_mlp_USPS = GridSearchCV(pipe_USPS_mlp_MinMaxScaler, param_grid=param_grid)
grid_MinMaxScaler_mlp_USPS.fit(X_train_USPS,y_train_USPS)
pred_USPS_MinMaxScaler_mlp=grid_MinMaxScaler_mlp_USPS.predict(X_test_USPS)
print("Test Error rate (MinMaxScaler) for classifier:", np.mean(y_test_USPS!=pred_USPS_MinMaxScaler_mlp))
print("Test Error score (MinMaxScaler) for wine dataset (SVC):", np.mean(y_test_USPS!=pred_USPS_MinMaxScaler))
print('\n')

#----- Normalizer ------------
pipe_USPS_Normalizer_mlp = make_pipeline(Normalizer(), mlp_USPS)
grid_Normalizer_mlp_USPS = GridSearchCV(pipe_USPS_Normalizer_mlp, param_grid=param_grid) 
grid_Normalizer_mlp_USPS.fit(X_train_USPS,y_train_USPS)
pred_USPS_Normalizer_mlp=grid_Normalizer_mlp_USPS.predict(X_test_USPS)
print("Test Error rate (Normalizer) for classifier on wine dataset:", 
      np.mean(y_test_USPS!=pred_USPS_Normalizer_mlp))
print("Test Error rate (Normalizer) for wine dataset:",np.mean(y_test_USPS!=pred_USPS_Normalizer))

Generalization accuracy (train) with MLP Classifier: 0.9925427670744421
Accuracy (training set) of the SVM on USPS dataset (default parameters): 0.9886705655186029
Test Error Rate (default parameters) with MLP Classifier: 0.008602150537634357
Test Error Rate of the SVM: 0.010322580645161339


Test Error rate (MinMaxScaler) for classifier: 0.004731182795698925
Test Error score (MinMaxScaler) for wine dataset (SVC): 0.006021505376344086


Test Error rate (Normalizer) for classifier on wine dataset: 0.0064516129032258064
Test Error rate (Normalizer) for wine dataset: 0.006881720430107527


Observations:

Question 4: 
It was observed that the result obtained for Accuracy on training set and test error rate on test dataset for both the datasets (Wine and USPS) are almost similar meaning that the model is underfitting where training and test performance is quite similar. The model can be improved by increasing the value of C and gamma.

Question 6:

Normalization techiques used on Wine Data set were StandardScaler, MinMaxScaler, RobustScaler, and Normalizer.

Wine Data

After training the dataset and testing it on different normalisation techniques, it was observed that for:

MinMaxScaler: 0.9924812030075187 (training) and 0.9777777777777777 (test score) - The performance decreased.

StandardScaler: 0.9924812030075187 (training) and 1.0 (test score) - The model overfitted

RobustScaler: 0.9849624060150376 (training) and 0.9777777777777777 (test score) - The model performace decreased

Normalizer: 0.9097744360902256 (training) and 0.9777777777777777 (test score) - The performace of the model increased.

Hence, for wine dataset Normalizer is a better normalization technique.


USPS Dataset

After training the dataset and testing it on different normalisation techniques, it was observed that for:

MinMaxScaler: 0.9939767675319088 (training) and 0.993978494623656 (test score) - The performance is underfitting.

StandardScaler: 0.9954108705005019 (training) and 0.9969892473118279 (test score) - The model performance increased.

RobustScaler: 0.9870930732826617 (training) and 0.9862365591397849 (test score) - The model performace decreased

Normalizer: 0.9929728954538936 (training) and 0.9931182795698925 (test score) - The performace of the model is almost the same for training and test - underfitting.

Hence, for USPS dataset StandardScaler is a better normalization technique.